File: app/vectorbt_test/core/nodes.py
from abc import ABC, abstractmethod
import enum
import pandas as pd
from vectorbt_test.core.context import PortfolioContext
from vectorbt_test.engine.execution_engine import Scope


class NodeType(enum.Enum):
    Unknown = "unknown"
    Indicator = "indicator"
    Factor = "factor"
    Signal = "signal"
    Function = "function"
    Data = "data"


class NodeDType(enum.Enum):
    ANY = "any"

    # 数值类
    NUMERIC = "numeric"     # Series (float/int)

    # 布尔类
    BOOL = "bool"           # 普通布尔（中间态）
    SIGNAL = "signal"       # 交易信号（最终态）

    # 结构类（你缺的）
    FRAME = "frame"         # DataFrame


class Node(ABC):
    _name: str = ""
    scope: Scope | None = None
    dtype: NodeDType | None = None

    def __init__(self, name: str | None = None, type: NodeType = NodeType.Unknown):
        self._type = type
        self._name = name or self.__class__.__name__

    def _args(self):
        return [self.name, self._type.value, self.dtype.value, self.scope]
    
    # ===== cache key =====
    def cache_key(self):
        return (self.__class__.__name__, tuple(self._args()))
    
    def apply(self, series, func, context: PortfolioContext):
        return context.execution_engine.apply(series, func, self.scope)

    # ===== 类型 =====
    @property
    def type(self):
        return self._type

    @property
    def name(self):
        return self._name

    @property
    def is_numeric(self):
        return self.dtype == NodeDType.NUMERIC

    @property
    def is_bool(self):
        return self.dtype == NodeDType.BOOL

    @property
    def is_signal(self):
        return self.dtype == NodeDType.SIGNAL

    @property
    def is_frame(self):
        return self.dtype == NodeDType.FRAME

    # ===== evaluate（统一缓存入口）=====
    def evaluate(self, data, context: PortfolioContext = PortfolioContext(), return_result=False):
        assert context.data_provider is not None
        return context.data_provider.get(self, data, context)

    @abstractmethod
    def compute(self, data: pd.DataFrame, context: PortfolioContext):
        pass

    # =========================
    # DSL：算术
    # =========================
    def __add__(self, other):
        return BinaryOp(self, to_node(other), "add")

    def __sub__(self, other):
        return BinaryOp(self, to_node(other), "sub")

    def __mul__(self, other):
        return BinaryOp(self, to_node(other), "mul")

    def __truediv__(self, other):
        return BinaryOp(self, to_node(other), "div")

    # =========================
    # DSL：比较
    # =========================
    def __gt__(self, other):
        return BinaryOp(self, to_node(other), "gt")

    def __lt__(self, other):
        return BinaryOp(self, to_node(other), "lt")

    def __ge__(self, other):
        return BinaryOp(self, to_node(other), "ge")

    def __le__(self, other):
        return BinaryOp(self, to_node(other), "le")

    def __eq__(self, other):
        return BinaryOp(self, to_node(other), "eq")

    def __ne__(self, other):
        return BinaryOp(self, to_node(other), "ne")

    # =========================
    # DSL：布尔
    # =========================
    def __and__(self, other):
        return BinaryOp(self, to_node(other), "and")

    def __or__(self, other):
        return BinaryOp(self, to_node(other), "or")

    def __invert__(self):
        return UnaryOp(self, "not")

    # =========================
    # DSL：函数
    # =========================
    def slope(self):
        return Slope(self)


class FeatureNode(Node):
    def __init__(self, type=NodeType.Unknown):
        super().__init__(type=type)
    pass


class ConstNode(FeatureNode):
    dtype = NodeDType.NUMERIC

    def __init__(self, value):
        super().__init__(NodeType.Indicator)
        self.value = value

    def _args(self):
        return [self.value] + super()._args()

    def compute(self, data, context: PortfolioContext):
        any_series = next(iter(context.data_provider.get_cache().values()))
        return any_series * 0 + self.value


def to_node(x):
    if isinstance(x, Node):
        return x
    return ConstNode(x)


class ArgNode(ConstNode):
    dtype = NodeDType.ANY

    def __init__(self, value):
        super().__init__(value)
        
    def compute(self, data, context: PortfolioContext):
        raise NotImplementedError("DO NOT call this, the ArgNode does not have busisness usage, only for framework to handle argument.")


def to_args(x):
    if isinstance(x, enum.Enum):
        x = x.value
      
    if isinstance(x, str):
        return ArgNode(x)
    return to_node(x)


class BinaryOp(FeatureNode):
    def __init__(self, left, right, op):
        super().__init__()

        self.left = left
        self.right = right
        self.op = op
    
        self.dtype = self._infer_dtype()
        self.scope = self._infer_scope()
    
    def _args(self):
        return [self.left.cache_key(), self.right.cache_key(), self.op] + super()._args()
    
    def _infer_dtype(self):
        l = self.left.dtype
        r = self.right.dtype

        # ===== 数值运算 =====
        if self.op in {"add", "sub", "mul", "div"}:
            if l != NodeDType.NUMERIC or r != NodeDType.NUMERIC:
                raise TypeError("Arithmetic requires numeric inputs")
            return NodeDType.NUMERIC

        # ===== 比较 =====
        if self.op in {"gt", "lt", "ge", "le", "eq", "ne"}:
            return NodeDType.BOOL

        # ===== 布尔 =====
        if self.op in {"and", "or"}:
            if l not in {NodeDType.BOOL, NodeDType.SIGNAL}:
                raise TypeError("AND requires bool/signal")
            if r not in {NodeDType.BOOL, NodeDType.SIGNAL}:
                raise TypeError("AND requires bool/signal")

            return NodeDType.BOOL

        raise ValueError(self.op)

    def _infer_scope(self):
        if self.left.scope == Scope.CS or self.right.scope == Scope.CS:
            return Scope.CS
        return Scope.TS

    def compute(self, data, context: PortfolioContext):
        l = self.left.evaluate(data, context)
        r = self.right.evaluate(data, context)

        if self.op == "add":
            return l + r
        elif self.op == "sub":
            return l - r
        elif self.op == "mul":
            return l * r
        elif self.op == "div":
            return l / r.replace(0, 1e-9)

        elif self.op == "gt":
            return (l > r).astype(int)
        elif self.op == "lt":
            return (l < r).astype(int)
        elif self.op == "ge":
            return (l >= r).astype(int)
        elif self.op == "le":
            return (l <= r).astype(int)
        elif self.op == "eq":
            return (l == r).astype(int)
        elif self.op == "ne":
            return (l != r).astype(int)

        elif self.op == "and":
            return (l.astype(bool) & r.astype(bool)).astype(int)
        elif self.op == "or":
            return (l.astype(bool) | r.astype(bool)).astype(int)

        raise ValueError(f"Unknown op {self.op}")
   

class UnaryOp(FeatureNode):
    dtype = NodeDType.BOOL

    def __init__(self, node, op):
        super().__init__(NodeType.Factor)
        self.node = node
        self.op = op

    def _args(self):
        return [self.node.cache_key(), self.op] + super()._args()

    def compute(self, data, context: PortfolioContext):
        x = self.node.evaluate(data, context)

        if self.op == "not":
            return (~x.astype(bool)).astype(int)

        raise ValueError(self.op)
   

class Slope(FeatureNode):
    dtype = NodeDType.NUMERIC

    def __init__(self, node):
        super().__init__(node.type)
        self.node = node

    def _args(self):
        return [self.node.cache_key()] + super()._args()

    def compute(self, data, context: PortfolioContext):
        x = self.node.evaluate(data, context)
        return x.diff()

File: quant-mantis/src/model/dsl_node/nodes.ts
import { NodeType, NodeDType, Scope, PortfolioContext } from "./types";

// 核心枚举和基础类型
// export enum NodeType {
//   Factor = "Factor",
//   Data = "Data",
//   Indicator = "Indicator",
//   Signal = "Signal",
//   Function = "Function"
// }

// export enum NodeDType {
//   NUMERIC = "NUMERIC",
//   BOOL = "BOOL",
//   SIGNAL = "SIGNAL",
//   ANY = "ANY",
//   FRAME = "FRAME",
//   NULL = "NULL"
// }

// export enum Scope {
//   TS = "TS",  // Time Series
//   CS = "CS"   // Cross Section
// }

// // ==========================
// // 上下文（仅占位，不实现）
// // ==========================
// export class PortfolioContext { }
// export class ExecutionEngine { }
// export class DataProvider { }

// ==========================
// Node 基类
// ==========================
export abstract class Node {
    _name: string;
    _type: NodeType;
    scope?: Scope;
    dtype?: NodeDType;

    constructor(name?: string, type: NodeType = NodeType.Unknown) {
        this._type = type;
        this._name = name || this.constructor.name;
    }

    _args(): any[] {
        return [this.name, this._type, this.dtype, this.scope];
    }

    cacheKey(): string {
        return JSON.stringify([this.constructor.name, ...this._args()]);
    }

    // ===== 类型判断 =====
    get type(): NodeType {
        return this._type;
    }

    get name(): string {
        return this._name;
    }

    get is_numeric(): boolean {
        return this.dtype === NodeDType.NUMERIC;
    }

    get is_bool(): boolean {
        return this.dtype === NodeDType.BOOL;
    }

    get is_signal(): boolean {
        return this.dtype === NodeDType.SIGNAL;
    }

    get is_frame(): boolean {
        return this.dtype === NodeDType.FRAME;
    }

    evaluate(data: any, context: PortfolioContext = new PortfolioContext(), return_result = false): any {
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        return_result;

        return this.compute(data, context);
    }

    compute(data: any, context: PortfolioContext): any {
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        data;
        context;
        throw new Error("Not implemented");
    }

    // =========================
    // DSL 操作符重载
    // =========================
    __add__(other: any): Node {
        return new BinaryOp(this, toNode(other), "add");
    }

    add(other: any): Node {
        return this.__add__(other);
    }

    __sub__(other: any): Node {
        return new BinaryOp(this, toNode(other), "sub");
    }

    sub(other: any): Node {
        return this.__sub__(other);
    }

    __mul__(other: any): Node {
        return new BinaryOp(this, toNode(other), "mul");
    }

    mul(other: any): Node {
        return this.__mul__(other);
    }

    __truediv__(other: any): Node {
        return new BinaryOp(this, toNode(other), "div");
    }

    div(other: any): Node {
        return this.__truediv__(other);
    }

    __gt__(other: any): Node {
        return new BinaryOp(this, toNode(other), "gt");
    }

    gt(other: any): Node {
        return this.__gt__(other);
    }

    __lt__(other: any): Node {
        return new BinaryOp(this, toNode(other), "lt");
    }

    lt(other: any): Node {
        return this.__lt__(other);
    }

    __ge__(other: any): Node {
        return new BinaryOp(this, toNode(other), "ge");
    }

    ge(other: any): Node {
        return this.__ge__(other);
    }

    __le__(other: any): Node {
        return new BinaryOp(this, toNode(other), "le");
    }

    le(other: any): Node {
        return this.__le__(other);
    }

    __eq__(other: any): Node {
        return new BinaryOp(this, toNode(other), "eq");
    }

    eq(other: any): Node {
        return this.__eq__(other);
    }

    __ne__(other: any): Node {
        return new BinaryOp(this, toNode(other), "ne");
    }

    ne(other: any): Node {
        return this.__ne__(other);
    }

    __and__(other: any): Node {
        return new BinaryOp(this, toNode(other), "and");
    }

    and(other: any): Node {
        return this.__and__(other);
    }

    __or__(other: any): Node {
        return new BinaryOp(this, toNode(other), "or");
    }

    or(other: any): Node {
        return this.__or__(other);
    }

    __invert__(): Node {
        return new UnaryOp(this, "not");
    }

    invert(): Node {
        return this.__invert__();
    }
    
    // 扩展方法
    slope(): Node {
        return new Slope(this);
    }
}

// =========================
// FeatureNode
// =========================
export class FeatureNode extends Node {
    constructor(type: NodeType = NodeType.Unknown) {
        super(undefined, type);
    }
}

// =========================
// 常量节点
// =========================
export class ConstNode extends FeatureNode {
    override dtype = NodeDType.NUMERIC;
    value: any;

    constructor(value: any) {
        super(NodeType.Indicator);
        this.value = value;
    }

    _args(): any[] {
        return [this.value, ...super._args()];
    }
}

// =========================
// to_node
// =========================
export function toNode(x: any): Node {
    if (x instanceof Node) return x;
    return new ConstNode(x);
}

// =========================
// ArgNode
// =========================
export class ArgNode extends ConstNode {
    override dtype = NodeDType.ANY;
    constructor(value: any) {
        super(value);
    }
}

// =========================
// to_args
// =========================
export function to_args(x: any): Node {
    if (typeof x === "object" && x?.value !== undefined) {
        x = x.value;
    }
    if (typeof x === "string") {
        return new ArgNode(x);
    }
    return toNode(x);
}

// =========================
// BinaryOp
// =========================
export class BinaryOp extends FeatureNode {
    left: Node;
    right: Node;
    op: string;

    constructor(left: Node, right: Node, op: string) {
        super();
        this.left = left;
        this.right = right;
        this.op = op;
        this.dtype = this._infer_dtype();
        this.scope = this._infer_scope();
    }

    _args(): any[] {
        return [this.left.cacheKey(), this.right.cacheKey(), this.op, ...super._args()];
    }

    _infer_dtype(): NodeDType {
        // @ts-ignore
        const l = this.left.dtype!;
        // @ts-ignore
        const r = this.right.dtype!;

        if (["add", "sub", "mul", "div"].includes(this.op)) {
            return NodeDType.NUMERIC;
        }

        if (["gt", "lt", "ge", "le", "eq", "ne"].includes(this.op)) {
            return NodeDType.BOOL;
        }

        if (["and", "or"].includes(this.op)) {
            return NodeDType.BOOL;
        }

        throw new Error("unknown op: " + this.op);
    }

    _infer_scope(): Scope {
        if (this.left.scope === Scope.CS || this.right.scope === Scope.CS) {
            return Scope.CS;
        }
        return Scope.TS;
    }
}

// =========================
// UnaryOp
// =========================
export class UnaryOp extends FeatureNode {
    override dtype = NodeDType.BOOL;
    node: Node;
    op: string;

    constructor(node: Node, op: string) {
        super(NodeType.Factor);
        this.node = node;
        this.op = op;
    }

    _args(): any[] {
        return [this.node.cacheKey(), this.op, ...super._args()];
    }
}

// =========================
// Slope（占位）
// =========================
export class Slope extends FeatureNode {
    constructor(node: Node) {
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        node;
        super();
    }
}